app/models/audit/event/authn.rb

```diff
@@ -0,0 +1,53 @@
+require 'ostruct'
+
+class Audit::Event::Authn < OpenStruct
```

Comment:
Currently the interface for this class is only implicit: you have to read through every method to figure out what attribute assignments must be made before it can be used.  It must be made explicit one way or the other.  My preference would be:
1. Write a normal intializer which accepts the member variables.  This has the advantage of not only being explicit, but also making the client code interface clear and consistent.  There's really no reason anyone should be using attr setters with this object.  Also, you can enforce immutability.
2. If you don't want to do 1. right now,  please at least document the intended interface in comments.

Reply:
Good point, I'll make it clearer.